### ISSUE TITLE
400 Bad Request errors component_id=loki

### ISSUE THREAD

**Issue by babykart** [vectordotdev/vector#19092]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

After upgrading to vector-0.34.0, the loki sink is in error :

```
2023-11-08T12:08:04.505443Z ERROR sink{component_kind="sink" component_id=loki component_type=loki}:request{request_id=29940}: vector::sinks::util::retries: Internal log [Non-retriable error; dropping the request.] is being suppressed to avoid flooding.
2023-11-08T12:08:04.505480Z ERROR sink{component_kind="sink" component_id=loki component_type=loki}:request{request_id=29940}: vector_common::internal_event::service: Internal log [Service call failed. No retries or retries exhausted.] is being suppressed to avoid flooding.
2023-11-08T12:08:04.505498Z ERROR sink{component_kind="sink" component_id=loki component_type=loki}:request{request_id=29940}: vector_common::internal_event::component_events_dropped: Internal log [Events dropped] is being suppressed to avoid flooding.
2023-11-08T12:08:14.834425Z ERROR sink{component_kind="sink" component_id=loki component_type=loki}:request{request_id=29945}: vector::sinks::util::retries: Internal log [Non-retriable error; dropping the request.] has been suppressed 5 times.
2023-11-08T12:08:14.834445Z ERROR sink{component_kind="sink" component_id=loki component_type=loki}:request{request_id=29945}: vector::sinks::util::retries: Non-retriable error; dropping the request. error=Server responded with an error: 400 Bad Request internal_log_rate_limit=true
2023-11-08T12:08:14.834469Z ERROR sink{component_kind="sink" component_id=loki component_type=loki}:request{request_id=29945}: vector_common::internal_event::service: Internal log [Service call failed. No retries or retries exhausted.] has been suppressed 5 times.
2023-11-08T12:08:14.834474Z ERROR sink{component_kind="sink" component_id=loki component_type=loki}:request{request_id=29945}: vector_common::internal_event::service: Service call failed. No retries or retries exhausted. error=Some(ServerError { code: 400 }) request_id=29945 error_type="request_failed" stage="sending" internal_log_rate_limit=true
```

Downgrade to vector-0.33.1 fix the problem.

### Configuration

```text
api:
  address: 0.0.0.0:8686
  enabled: true
  playground: false
data_dir: /var/lib/vector
sinks:
  loki:
    buffer:
      max_size: 268435488
      type: disk
    compression: snappy
    encoding:
      codec: json
    endpoint: http://loki-write.loki.svc.cluster.local:3100
    healthcheck: true
    inputs:
    - remap_kube
    labels:
      container_name: '{{ container_name }}'
      namespace: '{{ namespace }}'
      node_name: '{{ node_name }}'
      pod_name: '{{ pod_name }}'
    out_of_order_action: accept
    request:
      concurrency: adaptive
    tenant_id: kube
    tls:
      verify_certificate: false
      verify_hostname: false
    type: loki
sources:
  kube:
    extra_namespace_label_selector: kubernetes.io/metadata.name!=argocd,kubernetes.io/metadata.name!=harbor,kubernetes.io/metadata.name!=ingress,kubernetes.io/metadata.name!=loki
    type: kubernetes_logs
    use_apiserver_cache: true
transforms:
  remap_kube:
    inputs:
    - kube
    source: |
      .node_name = del(.kubernetes.pod_node_name)
      .namespace = del(.kubernetes.pod_namespace)
      .pod_name = del(.kubernetes.pod_name)
      .container_name = del(.kubernetes.container_name)
      del(.file)
      del(.kubernetes)
    type: remap
```

### Version

vector 0.34.0 (x86_64-unknown-linux-gnu c909b66 2023-11-07 15:07:26.748571656)

**Comment by precisionconage** [comment]
Also experienced this. It looks like Vector is sending the header as `application/json` rather than `application/x-protobuf` but still encoding in protobuf from what I can tell.

**Comment by LibiKorol** [comment]
once upgraded to vector ver 0.34.0 we received the same errors on loki sink as described above

**Comment by StephenWakely** [comment]
You should be able to fix this by removing the configuration option:

```yml
    compression: snappy
```

`snappy` is the default compression for this sink, so it will still have  the same behaviour for you. 

A bug crept in to v0.34.0 whereby the sink had two snappy config options and by specifying it in the configuration it is choosing the wrong one.

This will be fixed by #19099.

Let me know if removing that option does not fix things for you.

**Comment by babykart** [comment]
@StephenWakely I confirm : removing `compression: snappy` with vector-0.34.0 on the loki sink fix the error.
Thank you for your help and this explanation.

**Comment by jszwedko** [comment]
Closing since this is resolved in `master`. It'll be included in the `v0.34.1` release.
